packages/@lwc/style-compiler/src/serialize.ts

```diff
@@ -174,34 +174,46 @@ function serializeCss(result: Result, collectVarFunctions: boolean): string {
     return generateExpressionFromTokens(tokens);
 }
 
-// TODO [#1288]: this code needs refactor, it could be simpler by using a native post-css walker
-function tokenizeCssSelector(data: string): Token[] {
+// Given any CSS string, replace the scope tokens from the CSS with code to properly
+// replace it in the stylesheet function.
+function tokenizeCss(data: string): Token[] {
     data = data.replace(/( {2,})/gm, ' '); // remove when there are more than two spaces
-    const tokens: Token[] = [];
-    let pos = 0;
-    let next = 0;
-    const max = data.length;
 
-    while (pos < max) {
-        if (data.indexOf(`[${HOST_ATTRIBUTE}]`, pos) === pos) {
-            tokens.push({ type: TokenType.identifier, value: HOST_SELECTOR_IDENTIFIER });
+    const tokens: Token[] = [];
+    const regex = new RegExp(`[[-](${SHADOW_ATTRIBUTE}|${HOST_ATTRIBUTE})]?`, 'g');
 
-            next += HOST_ATTRIBUTE.length + 2;
-        } else if (data.indexOf(`[${SHADOW_ATTRIBUTE}]`, pos) === pos) {
-            tokens.push({ type: TokenType.identifier, value: SHADOW_SELECTOR_IDENTIFIER });
+    let lastIndex = 0;
+    for (const match of matchAll(data, regex)) {
+        const index = match.index!;
+        const [matchString, substring] = match;
 
-            next += SHADOW_ATTRIBUTE.length + 2;
-        } else {
-            next += 1;
+        if (index > lastIndex) {
+            tokens.push({ type: TokenType.text, value: data.substring(lastIndex, index) });
+        }
 
-            while (data.charAt(next) !== '[' && next < max) {
-                next++;
-            }
+        const identifier =
+            substring === SHADOW_ATTRIBUTE ? SHADOW_SELECTOR_IDENTIFIER : HOST_SELECTOR_IDENTIFIER;
```

Comment:
These variables should probably be renamed since they're not just attributes anymore, but I wanted to keep this PR small and focused.